src/python/pants/engine/legacy/graph.py

```diff
@@ -354,11 +356,21 @@ def hydrate_target(target_adaptor, hydrated_fields):
                         tuple(target_adaptor.dependencies))
 
 
-def _eager_fileset_with_spec(spec_path, filespec, snapshot, include_dirs=False):
+def _eager_fileset_with_spec(spec_path, filespec, base_globs, snapshot, kwarg_name,
+                             glob_match_failure, include_dirs=False, target_address=None):
   fds = snapshot.path_stats if include_dirs else snapshot.files
   files = tuple(fast_relpath(fd.path, spec_path) for fd in fds)
 
-  relpath_adjusted_filespec = FilesetRelPathWrapper.to_filespec(filespec['globs'], spec_path)
+  rel_include_globs = filespec['globs']
+
+  _get_globs_owning_files(files,
```

Comment:
Eek... sorry. Unfortunately, I think this is the wrong place to do this check. At this point, it's necessary to reverse engineer the match.

----

In the ticket, I said:
```
...since we now control the implementation of glob matching, we no longer have a great excuse not to.
```
...by which I meant: we have an implementation of glob matching in: https://github.com/pantsbuild/pants/blob/cb10df5444fcf28f39dbce3adf6ed22f0effc5d8/src/rust/engine/fs/src/lib.rs#L666

What I had in my head when I said that (but didn't explain very well: sorry), was that in `fn expand` we could track the "sources" of each `PathGlob` instance (ie, this PathGlob came from this original `glob` string), and then at the end see which original glob string never matched anything. I don't know concretely what this would look like, but I suspect that 1) it would be more efficient than re-constructing later, 2) it would be less likely to break.